### ISSUE TITLE
[Discussion] Issues with new code commenting style

### ISSUE THREAD

**Issue by initramfs** [qTox/qTox#3559]
Apologies for posting this here rather than in the [comment style change PR](https://github.com/tux3/qTox/pull/3536), I didn't see it until a few hours ago when it was already merged.

I'm in the processing of rebasing my code to the style introduced in PR #3536 when I found a few issues with the new style. First of all I would like to state that evaluating pros/cons, I agree with putting comments in the source files for **most cases**, but not all.

In particular, I would like to make the argument against putting comments for class descriptions (e.g. comments tagged with `@class`) and comments for variable descriptions (e.g. comments tagged with `@var`) in source files.

First of all, class comments basically declare the outline of a class, the functionality and the intended purpose or role for the class. These concepts usually do not change very often making header-triggered recompilation times minimal compared to other changes. Furthermore, it is very likely that a documentation change for the class changes at least some part of the header as well (whether this is a new function, refactoring of existing functions or the change in the parameters of certain functions) which would result in header-trigged recompliation regardless of where the documentation is placed.

By the change in PR #3536 class documentations just kinda float around the top of source files, making getting to them harder (functions are fine since QtCreator has a "Follow Symbol Under Cursor" option that takes you between the source and header for the function, allowing easy preview). Additionally, in the case of nested classes or multiple classes in one header, things get weird (do you put both class definitions at the top of the file? Or do you put the second/nested class documentation at the start of it's function implementions/constructors).

I've mentioned nested classes specifically because in one of my PRs I have a case of a nested class which is completely delared in the header (no functions declared in source). Do I isolate the class documention from the rest of the class and put it in the source file? That seems kinda strange.

Moving on to variable documentation, I believe that member variable documention must be kept right at the point of declaration _because_ of it's volatility. The fact that members are separated from their documentation, in my opinion, will only lead to changes in name/usage/type/role of the member variable that becomes inconsistent with the documentation due to failure to update documentation (people might not notice there is documentation for the particular variable for example). The same idea applies to nested/multiple classes, where do those variable documentations go?

Again, I made this point when I was rebasing my PR to the changes in master, having to cross-reference repeatedly between the header and source whether a particular variable was changed, renamed or deleted (within my PR).

A third but minor point regards the style of commenting used. From the codebase I've seen:

```
/**
 * Comment
 */
```

and

```
/// Comment
```

styles being used, in which the former is what QtCreator defaults to (at least on my system). The style of comment implemented by the PR is:

```
/**
Comment
*/
```

Which can easily be converted into the first style. Using the first style at least enforces consistency as developers won't have to worry about deliberately changing the comment style, having QtCreator automatically generate a valid template that's consistent with rest of the comments in the codebase.

Besides, even after the PR, I still see many classes who's comments have not been affected by the PR which use the first style mentioned above. Wouldn't just sticking to the first style allow for better integration? At least commenting styles would be consistent across the codebase whilst comment content can slowly be improved/added.

I've only brought these issues up now because comment consistency is something I deem extremely important, and I want to change my existing PR code to satisfy the "qTox" comment style (whether I agree with it or not). But before doing so, want to share a bit of my rationale regarding appropriate comment style.

---

**tl;dr** I believe `@class` and `@var` comments should be in headers and we should use the:

```
/**
 * Comment
*/
```

style of comments.

@Diadlo @sudden6 @zetok

**Comment by Diadlo** [comment]
https://github.com/tux3/qTox/pull/3488#discussion_r70524718
cc @antis81

**Edit**: I agree with your point of view. And I regret, that discussion about comment style was in comment to another PR, but not on in the separate issue

**Comment by antis81** [comment]
First, it is good to have clear and practical rules for commenting (API) code. Up to now I worked with several C++ based projects and the three possible "basic docs strategies":
1. Mixed (aka "I don't care")
   Problem with this is: Very often, people simply forget (or don't care) to update the docs and in most cases this is as good as no docs at all!
2. Put everything in Header:
   Well… that's not possible at all. So sooner or later this will become the same as point 1… :)
3. Put everything in Implementation:
   Keeping the docs completely in cpp files is actually good practice and I don't see any reason to leave that path. By putting the `@class, @enum, etc.` docs in the top section of the file, it also doesn't float around and is relatively easy to write/update. It also helps one to quickly get an overview of the class, because the header is lean and the important things of the class are documented in a single comment block (including enums, typedefs, etc. inside the class)!

The only good reason to put docs into header is, when there is no cpp available (e.g. in pure template implementations). Those are very rare though in qTox (tending strongly to non-existent).

For the "leading stars", I do not have a strong opinion on that. I just personally tend to leave them out, because they add risk of ambiguous text flow, which may confuse the parser and include an unwanted star.

@initramfs IMO we now have a rather practical guide and that's fine. Would you accept that? Please keep in mind, that with changing back and forth, we bind resources and I don't think it's worth it. I'd suggest rather to discuss [this one](https://github.com/tux3/qTox/pull/3524)… :)

**Comment by Diadlo** [comment]
> Additionally, in the case of nested classes or multiple classes in one header, things get weird (do you put both class definitions at the top of the file? Or do you put the second/nested class documentation at the start of it's function implementions/constructors).

@initramfs I think, that it should be on the top of cpp file.

> I have a case of a nested class which is completely delared in the header

If exist real reason to store class implementation in .h file (abstract class or template, etc.), you should put it in cpp file

**Comment by initramfs** [comment]
@antis81 Regarding:

> Put everything in Implementation: 
> Keeping the docs completely in cpp files is actually good practice and I don't see any reason to leave that path.

Part of the reason I opened this issue in the first place is to give reasons where that's not possible or doesn't make sense. Like I mentioned, a nested class that resides purely in the header (either a template class or small class with inline functions) would technically allow you to place it's class documentation inside the enclosing source file though that just makes things harder to read (since the bulk of the class is in the header).

> It also helps one to quickly get an overview of the class, because the header is lean

I would argue that a class doc at the start of a class doesn't really make the header less readable (functions and members are still bunched together), but that's my opinion.

> The only good reason to put docs into header is, when there is no cpp available (e.g. in pure template implementations). Those are very rare though in qTox (tending strongly to non-existent).

It just so happens that I'm working on a complete redesign of the video subsystem, much like your Audio API re-implementation. I plan to heavily use OO principles for modularity, including the creation of quite a few essential abstract base classes. Even if I weren't working on this, it doesn't harm to future proof the comment scheme to account future code changes to use C++ design constructs, rather than just dismiss something because "we don't current use it".

To be fair, even with a scheme that's enforced, it's not likely that documentation will ever be completely in the source or completely in the header. Either due to the fact that people are lazy or that it's not possible by the C++ standard (templates, abstract classes, etc). Granted, it's true I personally prefer to have at least the class documentation in the header and may have bias in the matter, what I'm trying to suggest here is a compromise that allows ease of use (since this matters a lot in terms of how willing people are going to adhere to a commenting scheme) whilst keeping as much stuff in source as possible. Having at least class documentation in headers would get rid of a lot of ambiguities regarding where to place documentation (another example is, do we lump public/private members together or do we place them separately since there is no more distinction in the comments whether a particular nested class/interface/enum is public or protected or private).

What I'm suggesting here is indeed a "mixed" scheme but this is **not** "I don't care", a precise scheme is defined to what goes in the header and what goes in the source. It just so happens that the rule is "everything in source, except in these cases (class docs, var docs, templates, abstract classes, etc)". Again, I'm not trying to make a big fuss here, just trying to help solidify a comment scheme that can hopefully be applied to all present and future code. If you don't see merit in my argument, I will oblige to the scheme devised in @Diadlo's PR.

Regarding leading stars, I'll admit I find the ones without leading stars ugly.. :)
Though I do believe there is still a case in terms of how much effort you're expecting people to put when submitting a PR with new code/documentation (can you imagine going through the entire PR and adding line notes to every place their QtCreator auto-generated document format needs to stripped of it's stars?)

P.S. @Diadlo The specific class I mentioned isn't abstract, but is more of a tiny helper class with all functions declared inline (each function is tiny too) thus I elected it to keep it in the header.

**Comment by sudden6** [comment]
The easy things first, I think

```
/**
 * Comment
 */
```

should be the comments style.

Second I think putting all documentation into the source files is the better way, because simple rules are easier to follow and the documentation will be generated and thus separate anyway. Since most of the time writing the documentation is the hardest part, we shouldn't make it even harder by adding rules.

**Comment by antis81** [comment]
> …Since most of the time writing the documentation is the hardest part, we shouldn't make it even harder by adding rules.

As you state by yourself, we introduced a few and easy to follow rules - which btw. is the reason the docs are now worth to write. Like I wrote above, I don't have a strong meaning on the leading stars. If you like that better, just put them in… :)

**Comment by initramfs** [comment]
Okay, seems like the consensus is:

``` c++
/**
 * Comment
 */
```

for the commenting style and **as much documenation in source as possible**. I'll update my PRs to this format now and close the issue if no one else has any input.

As an example:

File `foo.h`:

``` c++
class Foo
{
public:
    Foo();
    Foo(int x);

    void publicBar();

    constexpr unsigned int memberFoo = 2;
private:
    int privateBar();

    class NestedFoo
    {
    public:
        NestedFoo();
        virtual ~NestedFoo();

        /**
         * @brief An example nested virtual function.
         */
        virtual void NestedFoobar();
    };
};
```

---

File `foo.cpp`:

``` c++
#include "foo.h"

/**
 * @class Foo
 * @brief Example class Foo.
 *
 * Longer description of class Foo.
 */

/**
 * @var memberFoo
 * @brief Example member variable for Foo.
 */

/**
 * @class NestedFoo
 * @brief Example nested class NestedFoo within Foo.
 */

/**
 * @brief Default constructor for foo.
 */
Foo::Foo()
{
    // Constructor implementation...
}

/**
 * @brief Yet another constructor for foo.
 *
 * @param x Example parameter x.
 */
Foo::Foo(int x)
{
    // Constructor implementation...
}

/**
 * @brief Example function publicBar for Foo.
 */
void Foo::publicBar()
{
    // Function implementation...
}

/**
 * @brief Example private function privateBar for Foo.
 *
 * @return The answer to life, the universe and everthing.
 */
int Foo::privateBar()
{
    // Function implementation...
    return 42;
}

/**
 * @brief Constructor for NestedFoo.
 */
NestedFoo::NestedFoo(){}

/**
 * @brief Destructor for NestedFoo.
 */
NestedFoo::~NestedFoo(){}
```

@antis81 @sudden6 @zetok @Diadlo
